File: backend/app/schemas/response.py
from typing import List, Literal, Union
from app.utils.enums import AgentType
from pydantic import BaseModel, Field
from uuid import uuid4


class Message(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid4()))
    msg_type: Literal["system", "agent"] | None  # system msg | agent messsage
    content: str | None = None


class SystemMessage(Message):
    msg_type: str = "system"
    type: Literal["info", "warning", "success", "error"] = "info"


class AgentMessage(Message):
    msg_type: str = "agent"
    agent_type: AgentType  # CoderAgent | WriterAgent


class CodeExecution(BaseModel):
    res_type: Literal["stdout", "stderr", "result", "error"]
    msg: str | None = None


class StdOutModel(CodeExecution):
    res_type: str = "stdout"


class StdErrModel(CodeExecution):
    res_type: str = "stderr"


class ResultModel(CodeExecution):
    res_type: str = "result"
    format: Literal[
        "text",
        "html",
        "markdown",
        "png",
        "jpeg",
        "svg",
        "pdf",
        "latex",
        "json",
        "javascript",
    ]


class ErrorModel(CodeExecution):
    res_type: str = "error"
    name: str
    value: str
    traceback: str


# 总返回类型
OutputItem = Union[StdOutModel, StdErrModel, ResultModel, ErrorModel]


# 1. 只带 code
# 2. 只带 code result
class CoderMessage(AgentMessage):
    agent_type: AgentType = AgentType.CODER
    code: str | None = None
    code_results: list[OutputItem] | None = None


class WriterMessage(AgentMessage):
    agent_type: AgentType = AgentType.WRITER

File: frontend/src/utils/response.ts
// 对应 response.py 的结构
export type SystemMessageType = 'info' | 'warning' | 'success' | 'error';
export type AgentType = 'CoderAgent' | 'WriterAgent';

export interface BaseMessage {
  id: string;
  msg_type: 'system' | 'agent' | 'user';
  content?: string;
}

export interface SystemMessage extends BaseMessage {
  msg_type: 'system';
  type: SystemMessageType;
}

export interface AgentMessage extends BaseMessage {
  msg_type: 'agent';
  agent_type: AgentType;
}

export interface CodeExecutionResult {
  res_type: string;
  msg: string;
}

export interface CoderMessage extends AgentMessage {
  agent_type: 'CoderAgent';
  code?: string;
  code_result?: CodeExecutionResult;
}

export interface WriterMessage extends AgentMessage {
  agent_type: 'WriterAgent';
}

export interface UserMessage extends BaseMessage {
  msg_type: 'user';
  content: string;
}

export type Message = SystemMessage | CoderMessage | WriterMessage | UserMessage